opt_einsum/contract.py

```diff
@@ -500,14 +500,35 @@ def contract(*operands, **kwargs):
     return _core_contract(operands, contraction_list, backend=backend, **einsum_kwargs)
 
 
-def _core_contract(operands, contraction_list, backend='numpy', evaluate_constants=False, **einsum_kwargs):
+def infer_backend(x):
+    return x.__class__.__module__.split('.')[0]
+
+
+def parse_backend(arrays, backend):
+    """Find out what backend we should use, dipatching based on the first
+    array if ``backend='auto'`` is specified.
+    """
+    if backend != 'auto':
+        return backend
+    backend = infer_backend(arrays[0])
+
+    # some arrays will be defined in modules that don't implement tensordot
+    # etc. so instead default to numpy
+    if not backends.has_tensordot(backend):
```

Comment:
What if someone has subclassed a NumPy array or similar or uses an unknown backend? We should fall back to NumPy here as well.

Reply:
So I _think_ the behavior you describe is what happens here. For example, in my library ``quimb`` I have a subclass of ``ndarray`` so the initial inferred backend is ``quimb``. However this check for ``quimb.tensordot`` (which is cached) means ``'numpy'`` is used still since unlike dask, tensorflow etc. ``quimb`` does not provide tensordot.